server/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java

```diff
@@ -290,32 +305,106 @@ protected AnalyzedRelation visitJoin(Join node, StatementAnalysisContext stateme
                         relationContext.parentSources(),
                         coordinatorTxnCtx.sessionSettings().searchPath().currentSchema()),
                     new SubqueryAnalyzer(this, statementContext));
-                Expression expr;
-                if (joinCriteria instanceof JoinOn) {
-                    expr = ((JoinOn) joinCriteria).getExpression();
-                } else {
-                    expr = JoinUsing.toExpression(
-                        leftRel.relationName().toQualifiedName(),
-                        rightRel.relationName().toQualifiedName(),
-                        ((JoinUsing) joinCriteria).getColumns());
+                var expressionAnalysisContext = statementContext.currentRelationContext().expressionAnalysisContext();
+                joinCondition = expressionAnalyzer.convert(expression, expressionAnalysisContext);
+            } catch (RelationUnknown e) {
+                throw new RelationValidationException(
+                    e.getTableIdents(),
+                    String.format(Locale.ENGLISH, "missing FROM-clause entry for relation '%s'", e.getTableIdents())
+                );
+            }
+        }
+        var joinType = JoinType.values()[node.getType().ordinal()];
```

Comment:
This looks a bit error prone as it assumes the two enums have the same elements in the same order. 

Could we merge the two or have a static factory that uses a switch expression to map from one to the other?

Reply:
Good idea.